core/blockchain_impl.go

```diff
@@ -1621,9 +1623,20 @@ func (bc *BlockChainImpl) InsertChain(chain types.Blocks, verifyHeaders bool) (i
 		return len(chain), nil
 	}
 
+	for _, b := range chain {
+		// check if blocks already exist
+		if bc.HasBlock(b.Hash(), b.NumberU64()) {
+			return 0, errors.Wrapf(ErrKnownBlock, "block %s %d already exists", b.Hash().Hex(), b.NumberU64())
```

Comment:
Have you look at all the method that called Insertchain and the potential impact on the protocol to return an error if the block already exist ? cc @GheisMohammadi 

also I assumed this happened in testnet for you because of the DB folder of the external validator that wasn't deleted, is that correct ?

Reply:
Also I am thinking that fix could be in a different PR so it deserve a proper code analysis of the impacted method called in the protocol logic